### PR TITLE
Add Koin dependency injection

### PR DESCRIPTION
**Pull request by Ch4t4r**
Currently the app has a lot of potential leaks where contexts are stored in static fields. This PR adds Koin to avoid this.

Things left to do:
 - ~~Fix remaining errors~~
 - ~~Check other classes for potential leaks~~
    - At least two leaks remain which would take way too much effort to fix.
 - ~~Move preferences to Koin?~~
    - Done

### CODE REVIEW

**Comment by Ch4t4r** [comment]
I haven't been able to actually run the app yet to verify that it still works.

Also, there are some instances where I'm not sure whether I've captured to logic correctly (e.g `GoVpnAdapter` with `appMode`), it would be great if you could check that.

**Comment by ignoramous** [comment]
cc: @hussainmohd-a

**Comment by ignoramous** [comment]
Thanks a lot for putting in your time into this. Appreciate it very much.

> At least two leaks remain which would take way too much effort to fix.

Where? Are they unfixable?

> ...there are some instances where I'm not sure whether I've captured to logic correctly

Hussain will cross verify in a day or two. This is his domain (and now, yours).

**Comment by Ch4t4r** [comment]
> Where? Are they unfixable?

1. `VpnController` has a static field `braveVpnService`. This is not an important fix because it is set to null with `stop()`. If the service is killed by the system it is pulled from memory entirely, so no reason to do cleanup.
2. `HomeScreenActivity` leaks `AppMode`, which references `Context` and multiple DAOs. It is used extensivly across the app. It is fixable, but I'm not sure whether I'd mess up the logic behind it.
3. `GoVpnAdapter` has a non-static `AsyncTask`. It potentially leaks Context through `this` as it is an inner class. It also is fixable, but requires refactor (also, IntelliSense doesn't work in this class for me)

After this has been merged `LeakCanary` could be added to check for further leaks: https://github.com/square/leakcanary

File: app/src/main/java/com/celzero/bravedns/receiver/ReceiverHelper.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.celzero.bravedns.receiver
```
Comment (ignoramous) [inline app/src/main/java/com/celzero/bravedns/receiver/ReceiverHelper.kt:1]: nit: missing license headers.

File: app/src/main/java/com/celzero/bravedns/receiver/ReceiverHelper.kt [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+import com.celzero.bravedns.service.PersistentState
+import org.koin.core.component.KoinComponent
+import org.koin.core.component.inject
+
+object ReceiverHelper:KoinComponent {
+    val persistentState by inject<PersistentState>()
+}
```
Comment (ignoramous) [inline app/src/main/java/com/celzero/bravedns/receiver/ReceiverHelper.kt:9]: minor: missing eof newline char `\n`

File: app/src/main/java/com/celzero/bravedns/receiver/BraveScreenStateReceiver.kt [hunk 3]
```diff
@@ -64,7 +66,7 @@ class BraveScreenStateReceiver : BroadcastReceiver() {
             val appInfoRepository = mDb.appInfoRepository()
             appInfoRepository.removeUninstalledPackage(pacakgeName)
             HomeScreenActivity.GlobalVariable.appList.remove(pacakgeName)
-            PersistentState.setExcludedPackagesWifi(packageName, true, context)
+            ReceiverHelper.persistentState.setExcludedPackagesWifi(packageName, true, context)
```
Comment (ignoramous) [inline app/src/main/java/com/celzero/bravedns/receiver/BraveScreenStateReceiver.kt:69]: Nice catch :) I think this entire code comment can be nuked.

File: app/src/main/java/com/celzero/bravedns/receiver/BraveScreenStateReceiver.kt [hunk 4]
```diff
@@ -32,19 +34,19 @@ class BraveScreenStateReceiver : BroadcastReceiver() {
...
                 newIntent.action = DeviceLockService.ACTION_CHECK_LOCK
                 newIntent.putExtra(DeviceLockService.EXTRA_STATE, intent.action)
-                context.startService(newIntent)
+                context!!.startService(newIntent)
             }
         } else if (intent.action.equals(Intent.ACTION_USER_PRESENT) || intent.action.equals(Intent.ACTION_SCREEN_ON)) {
             if(DEBUG) Log.d(LOG_TAG,"BraveScreenStateReceiver : ACTION_USER_PRESENT/ACTION_SCREEN_ON detected from the receiver")
-            if(PersistentState.getFirewallModeForScreenState(context!!)) {
-                val state = PersistentState.getScreenLockData(context)
+            if(ReceiverHelper.persistentState.getFirewallModeForScreenState()) {
```
Comment (ignoramous) [inline app/src/main/java/com/celzero/bravedns/receiver/BraveScreenStateReceiver.kt:46]: Is `ReceiverHelper` here a static singleton or an instance of type `ReceiverHelper`?

**Review by ignoramous — CHANGES_REQUESTED** [review]
Thanks for these massive changes! They look good to me. I guess you'd have to rebase against the main branch again, I hope that isn't painful.

Most comments are pointers to add apache license headers to new files and newline characters `\n` at  end of file... so, nothing major at all.

File: app/src/main/java/com/celzero/bravedns/ServiceModuleProvider.kt [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+val RootModule = module {
+    single<ContentResolver> { androidContext().contentResolver }
+}
+
+val AppModules:List<Module> by lazy {
+    mutableListOf<Module>().apply {
+        add(RootModule)
+    }
+}
```
[the remark this answers is not in the document]
Comment (Ch4t4r) [reply app/src/main/java/com/celzero/bravedns/ServiceModuleProvider.kt]: Done

File: app/src/main/java/com/celzero/bravedns/RethinkDnsApplication.kt [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+class RethinkDnsApplication:Application() {
+    override fun onCreate() {
+        super.onCreate()
+        startKoin {
+            if(BuildConfig.DEBUG) androidLogger()
+            androidContext(this@RethinkDnsApplication)
+            koin.loadModules(AppModules)
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (Ch4t4r) [reply app/src/main/java/com/celzero/bravedns/RethinkDnsApplication.kt]: Done

File: app/src/main/java/com/celzero/bravedns/RethinkDnsApplication.kt [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package com.celzero.bravedns
```
[the remark this answers is not in the document]
Comment (Ch4t4r) [reply app/src/main/java/com/celzero/bravedns/RethinkDnsApplication.kt:1]: Done

File: app/src/main/java/com/celzero/bravedns/ServiceModuleProvider.kt [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package com.celzero.bravedns
```
[the remark this answers is not in the document]
Comment (Ch4t4r) [reply app/src/main/java/com/celzero/bravedns/ServiceModuleProvider.kt:1]: Done

File: app/src/main/java/com/celzero/bravedns/database/DatabaseModule.kt [hunk 9]
```diff
@@ -0,0 +1,51 @@
+package com.celzero.bravedns.database
```
[the remark this answers is not in the document]
Comment (Ch4t4r) [reply app/src/main/java/com/celzero/bravedns/database/DatabaseModule.kt:1]: Done

File: app/src/main/java/com/celzero/bravedns/ui/ApplicationManagerActivity.kt [hunk 10]
```diff
@@ -115,41 +117,6 @@ class ApplicationManagerActivity : AppCompatActivity(), SearchView.OnQueryTextLi
         }
     }
 
-
-    companion object{
```
[the remark this answers is not in the document]
Comment (Ch4t4r) [reply app/src/main/java/com/celzero/bravedns/ui/ApplicationManagerActivity.kt:116]: Well, it wasn't used, hence I deleted it. If you want to share display logic you should create a custom View, a dedicated Fragment or use a shared ViewModel (although the latter wouldn't follow MVVM as nicely as you'd move display logic where it normally does not belong).
In this case a custom adapter class could also work.

File: app/src/main/java/com/celzero/bravedns/receiver/BraveScreenStateReceiver.kt (hunk 4 above)
Comment (Ch4t4r) [reply app/src/main/java/com/celzero/bravedns/receiver/BraveScreenStateReceiver.kt:46]: It is a Kotlin `object`, which is a singleton. It inherits from `KoinComponent` to be able to use `inject()` and provide the required services.

File: app/src/main/java/com/celzero/bravedns/database/DatabaseModule.kt [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+            RefreshDatabase(
+                androidContext(),
+                get(),
+                get(),
+                get(),
+                get(),
+                get(),
+                get(),
+                get(),
+                get()
```
[the remark this answers is not in the document]
Comment (Ch4t4r) [reply app/src/main/java/com/celzero/bravedns/database/DatabaseModule.kt:36]: Yes, `get<Any>()` is generic and retrieves the DAOs declared just below.

**Comment by Ch4t4r** [comment]
I just build the app and let it run on an emulator. It still builds, launches and properly forwards DNS queries -- I checked with a DNS leak test.
I obviously didn't check every settings, but tried clicking trough multiple views, but most certainly haven't gotten to all of them.

**Comment by ignoramous** [comment]
Rebased, squashed, and merged: https://github.com/celzero/rethink-app/commit/1238d4c4c05c26d6cd41d4de9c7c0f69e4c06157

Thanks! :D

**Pull request closed without merging**